drivers/platform/maxim/maxim_rtc.c

```diff
@@ -0,0 +1,315 @@
+/************************* Include Files **************************************/
+/******************************************************************************/
+
+#include <stdlib.h>
+#include <errno.h>
+#include "no-os/irq.h"
+#include "no-os/rtc.h"
+#include "no-os/util.h"
+#include "rtc.h"
+#include "rtc_extra.h"
+#include "rtc_regs.h"
+
+#define MS_TO_RSSA(x) (0 - ((x * 256) / 1000))
+
+/**
+* @brief Callback descriptor that contains a function to be called when an
+* interrupt occurs.
+*/
+static struct callback_desc *cb;
+
+/******************************************************************************/
+/************************ Functions Definitions *******************************/
+/******************************************************************************/
+
+void RTC_IRQHandler()
+{
+
+	mxc_rtc_regs_t *rtc_regs = MXC_RTC;
+	volatile uint32_t rtc_ctrl = rtc_regs->ctrl;
+	uint8_t n_int = 0;
+	/** Sub-second alarm flag clear */
+	rtc_regs->ctrl &= ~BIT(7);
+	/** Time-of-day alarm flag clear */
+	rtc_regs->ctrl &= ~BIT(6);
+	/** RTC (read) ready flag */
+	rtc_regs->ctrl &= ~BIT(5);
+
+	if(!cb)
+		return;
+
+	/** Shift right so the interrupt flags will be the first 3 bits */
+	rtc_ctrl >>= 5UL;
+	/** Clear the remaining bits */
+	rtc_ctrl &= 0x7UL;
+	while(rtc_ctrl) {
+		n_int = find_first_set_bit(rtc_ctrl);
+		if (rtc_ctrl & (rtc_regs->ctrl & BIT(n_int))) {
+			cb->callback(cb->ctx, n_int, cb->config);
+		}
+		rtc_ctrl >>= n_int + 1;
+	}
+}
+
+/**
+ * @brief Initialize the RTC peripheral.
+ * @param device - The RTC descriptor.
+ * @param init_param - The structure that contains the RTC initialization.
+ * @return 0 in case of success, errno codes otherwise.
+ */
+int32_t rtc_init(struct rtc_desc **device, struct rtc_init_param *init_param)
+{
+	int32_t ret = 0;
+	struct rtc_desc *dev;
+	sys_cfg_rtc_t sys_cfg;
+
+	if(!init_param)
+		return -EINVAL;
+
+	struct rtc_init_maxim *maxim_init_param = init_param->extra;
+	dev = calloc(1, sizeof(*dev));
+	if (!dev)
+		return -ENOMEM;
+
+	dev->id = init_param->id;
+	dev->freq = init_param->freq;
+	dev->load = init_param->load;
+
+	TMR_Enable(MXC_TMR0);
+
+	if (RTC_Init(MXC_RTC, dev->load, maxim_init_param->ms_load,
+		     &sys_cfg) != E_NO_ERROR) {
+		ret = -1;
+		goto error;
+	}
+
+	*device = dev;
+
+	return 0;
+
+error:
+	free(dev);
+
+	return ret;
+}
+
+/**
+ * @brief Free the resources allocated by rtc_init().
+ * @param dev - The RTC descriptor.
+ * @return 0 in case of success, errno codes otherwise.
+ */
+int32_t rtc_remove(struct rtc_desc *dev)
+{
+	if (!dev)
+		return -EINVAL;
+	rtc_unregister_callback();
+	free(dev);
+
+	return 0;
+}
+
+/**
+ * @brief Start the real time clock.
+ * @param dev - The RTC descriptor.
+ * @return 0 in case of success, errno codes otherwise.
+ */
+int32_t rtc_start(struct rtc_desc *dev)
+{
+	RTC_EnableRTCE(MXC_RTC);
+
+	/** Wait for synchronization */
+	if (RTC_CheckBusy())
+		return -EBUSY;
```

Comment:
nitpick: I think the comment is a bit misleading since it doesn't do any actual waiting, just returns an error if the RTC is busy.

Reply:
RTC_CheckBusy actually blocks the execution until the RTC's value can be read. It returns EBUSY if it waits for too long.